src/probnum/linops/_block.py

```diff
@@ -0,0 +1,134 @@
+"""Operators that utilize block structure."""
+from __future__ import annotations
+
+import functools
+
+import numpy as np
+from scipy.linalg import block_diag
+
+from probnum.typing import DTypeLike, LinearOperatorLike
+
+from . import _linear_operator, _utils
+
+
+class BlockDiagonalMatrix(_linear_operator.LinearOperator):
+    """Forms a block diagonal matrix from the input linear operators.
+
+    Given linear operators A, B, ..., Z, this represents the block linear
+    operator
+
+    .. math::
+        \\begin{bmatrix}
+        A & & & & \\\\
+        & B & & & \\\\
+        & & \\ddots & & \\\\
+        & & & \\ddots & \\\\
+        & & & & Z
+        \\end{bmatrix}
+
+    Parameters
+    ----------
+    blocks :
+        The (ordered!) input linear operators.
+    """
+
+    def __init__(self, *blocks: LinearOperatorLike):
+        blocks = tuple(_utils.aslinop(block) for block in blocks)
+        assert len(blocks) > 0
+
+        dtype = functools.reduce(np.promote_types, (block.dtype for block in blocks))
+        shape_0 = sum(block.shape[0] for block in blocks)
+        shape_1 = sum(block.shape[1] for block in blocks)
+        self.split_indices = np.array(
+            tuple(block.shape[1] for block in blocks)
+        ).cumsum()[:-1]
+
+        super().__init__((shape_0, shape_1), dtype)
+
+        self.is_symmetric = all(block.is_symmetric for block in blocks)
+        self.is_positive_definite = all(block.is_positive_definite for block in blocks)
+        self.is_upper_triangular = all(block.is_upper_triangular for block in blocks)
+        self.is_lower_triangular = all(block.is_lower_triangular for block in blocks)
```

Comment:
These assignments should adhere to the ternary logic used in the matrix properties.

For instance, if all the blocks are symmetric, then `self.is_symmetric` should be set to `True` and if any of the blocks is known non-symmetric (`block.is_symmetric is False`), then `self.is_symmetric` should be set to `False`. But if none of the blocks are known non-symmetric and there are some blocks for which `block.is_symmetric is None`, then `self.is_symmetric` should be set to `None`, since we don't know anything about its symmetry properties in this case.

The same holds for the other properties.

Reply:
Good catch, thanks!